roadmap/implementers-guide/src/node/approval/approval-voting.md

```diff
@@ -221,8 +223,8 @@ On receiving an `ApprovedAncestor(Hash, BlockNumber, response_channel)`:
   * Set `n_tranches = tranches_to_approve(approval_entry)`
   * If `OurAssignment` has tranche `<= n_tranches`, the tranche is live according to our local clock (based against block slot), and we have not triggered the assignment already
     * Import to `ApprovalEntry`
-    * Broadcast on network with an `ApprovalNetworkingMessage::DistributeAssignment`.
-    * Kick off approval work with `launch_approval`
+    * Broadcast on network with an `ApprovalDistributionMessage::DistributeAssignment`.
+    * Kick off approval work with `launch_approval`. Note that if the candidate appears in multiple current blocks, we will launch approval for each block it appears in. It may make sense to shortcut around this with caching either at this level or on the level of the other subsystems invoked by that function.
```

Comment:
Yes, these approval checkers being different is an unfortunate consequence of using the relay VRF modulo criteria for tranche zero.  

If the relay chain largely avoids forks then the relay VRF modulo criteria ensures all validators get the same number of tranche zero assignments, which improves tranche zero assignments in several respects:
- We avoid a choice between either overworking nodes so much or else giving them more arbitrary choices about what they check.  
- We estimate the number of checkers with a binimial distribution, which gives a variance n p q which looks less than the variance obtained from the more complex distributions resulting from other strategies.
- We could support validators of different computing capacities if we force validators to take assignments with sample numbers 1,..,k where k cannot exceed by too much their previous average number of samples completed on the same inclusion batch.  In this way, we pay more for validators with more CPU but permit others to participate.  This might not be a good thing, but nice option.

Reply:
All that sounds fine, but the point here is actually about the fact that we literally re-execute the candidate even if we've just validated it and we should have some LRU cache somewhere so that being "re-assigned" to the same candidate across multiple forks is ultra cheap.

Reply:
I missed interpreted and read this in terms of across validators, thanks!